### PR TITLE
:sparkles: Build multiarch images using buildx

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
**What this PR does / why we need it**:
- Build multiarch images using buildx

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
@Prajyot-Parab - is there an issue for this? I'm not fully clear on what problem this change is solving

**Comment by sbueringer** [comment]
@Prajyot-Parab Will take a look soon. Just fyi, just merged https://github.com/kubernetes-sigs/cluster-api/pull/8799. Would be good if we can adjust the targets for the new provider as well.

**Comment by Prajyot-Parab** [comment]
> @Prajyot-Parab Will take a look soon. Just fyi, just merged #8799. Would be good if we can adjust the targets for the new provider as well.

done.

**Comment by Prajyot-Parab** [comment]
> > ... leveraging the docker buildx capabilities to build multi-arch images via a single command.
> 
> Can you please point to where multi-arch images are being built via a single command?
> 
> AFAIK, the single command approach is to specify multiple arguments to `buildx build --platform`, e.g. `buildx build --platform linux/amd64,linux/arm64`. However, this fails with `--output=type=docker`. It appears to be designed for `--output=type=registry`, which means a multi-arch build requires a registry.
> 
> > This should also theoretically reduce the build time (as it supports parallel building, cache, etc)
> 
> Could you please explain, or link to documentation? It's not clear to me. (Is it because one buildx "builder" is used to build all the images?)

Ref - https://www.docker.com/blog/how-to-rapidly-build-multi-architecture-images-with-buildx/
Ref - https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/

Note - The multi-arch(manifest) push via buildx only supports registry, you can try individual arch image build commands to test with `-output=type=docker`

**Comment by fabriziopandini** [comment]
I'll defer to @sbueringer and @killianmuldoon, but it seems to me that the buildx approach doesn't simplify things in a significative way, and while we are adding a dependency on another tool + another set of settings on developers/ci machines 🤔

**Comment by Jont828** [comment]
@Prajyot-Parab Looks like you need to rebase btw.

File: hack/init-buildx.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+fi
+
+# Ensure qemu is in binfmt_misc
+# Docker desktop already has these in versions recent enough to have buildx
+# We only need to do this setup on linux hosts
+if [ "$(uname)" == 'Linux' ]; then
+  # NOTE: this is pinned to a digest for a reason!
+  # Running an image with the --privileged flag carries risks, while using a tag exposes you to trouble if a new malicious image is pushed
+  # However, using a digest ensures safety that is practically impossible to compromise
+  docker run --rm --privileged tonistiigi/binfmt:qemu-v7.0.0-28@sha256:66e11bea77a5ea9d6f0fe79b57cd2b189b5d15b93a2bdb925be22949232e4e55 --install all
```
Comment (mdbooth) [inline hack/init-buildx.sh:50]: What is this image?

File: hack/init-buildx.sh
Comment (mdbooth) [inline hack/init-buildx.sh:1]: Why is this required?

**Review by mdbooth — COMMENTED** [review]
The major benefit of switching to buildx is the move to multi-arch images. It means we can remove everything relating to single-arch images from the Makefile, and deployment templates become architecture independent because the container runtime just pulls the correct image automatically.

**Comment by sbueringer** [comment]
> The major benefit of switching to buildx is the move to multi-arch images. It means we can remove everything relating to single-arch images from the Makefile, and deployment templates become architecture independent because the container runtime just pulls the correct image automatically.

Sorry, can't find the time to dive deeper. Cluster API today already publishes multi-arch images. Are there different kind of multi-arch images?

**Comment by mdbooth** [comment]
> > The major benefit of switching to buildx is the move to multi-arch images. It means we can remove everything relating to single-arch images from the Makefile, and deployment templates become architecture independent because the container runtime just pulls the correct image automatically.
> 
> Sorry, can't find the time to dive deeper. Cluster API today already publishes multi-arch images. Are there different kind of multi-arch images?
> 
> ```shell
>  docker buildx imagetools inspect registry.k8s.io/cluster-api/cluster-api-controller:v1.5.0
> Name:      registry.k8s.io/cluster-api/cluster-api-controller:v1.5.0
> MediaType: application/vnd.docker.distribution.manifest.list.v2+json
> Digest:    sha256:f34016d3a494f9544a16137c9bba49d8756c574a0a1baf96257903409ef82f77
> 
> Manifests:
>   Name:      registry.k8s.io/cluster-api/cluster-api-controller:v1.5.0@sha256:c19cf5bb1a0c375eb0a67765bde1b9a8cff3707bcdbfee6a2023a13a8bab1c2b
>   MediaType: application/vnd.docker.distribution.manifest.v2+json
>   Platform:  linux/amd64
> 
>   Name:      registry.k8s.io/cluster-api/cluster-api-controller:v1.5.0@sha256:8fc66cc4944cacb07bb5c38bab3840ff2a3ff88241c2b3e6516da3d72478767b
>   MediaType: application/vnd.docker.distribution.manifest.v2+json
>   Platform:  linux/arm
> 
> ...
> ```

Ah, I'd forgotten we were doing this ourselves. buildx does it for you. I have found it simpler to use, understand, and maintain.

File: hack/init-buildx.sh (hunk 1 above)
Comment (vincepri) [reply hack/init-buildx.sh:50]: /hold

We should not ever run a privileged container somewhere.

File: hack/init-buildx.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+fi
+
+# Ensure we use a builder that can leverage it (the default on linux will not)
+docker buildx rm capi >/dev/null 2>&1 || true
+docker buildx create --use --name=capi
```
Comment (vincepri) [inline hack/init-buildx.sh:55]: Do we absolutely need to use `buildx` for this? Is there no other more open alternative solution?

**Review by vincepri — CHANGES_REQUESTED** [review]

**Comment by sbueringer** [comment]
@Prajyot-Parab Apologies for the huge delay in reviewing this. I know you mostly opened this PR because I asked you to, once you figured out how it works in your provider.

Looking at this PR as of now, I don't think it's worth introducing the additional complexity. As far as I can tell with both implementations we end up with the same multi-arch images. As of now I don't see the benefit in building the images with buildx compared to how we do it right now (please let me know If I'm missing something).
